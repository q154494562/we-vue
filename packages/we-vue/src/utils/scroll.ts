type ElementOrWindow = HTMLElement | Window

/**
 * 找到最近的触发滚动事件的元素
 * @param {Element} element
 * @param {Element} rootParent
 * @returns {Element | window}
 */
export function getScrollEventTarget (element: HTMLElement, rootParent: ElementOrWindow = window) {
  let node = element
  while (
    node &&
    node.tagName !== 'HTML' &&
    node.tagName !== 'BODY' &&
    node.nodeType === 1 &&
    node !== rootParent
  ) {
    const { overflowY } = window.getComputedStyle(node)
    if (overflowY === 'scroll' || overflowY === 'auto') {
      return node
    }
    node = <HTMLElement>node.parentNode
  }
  return rootParent
}

// 获取滚动高度
export function getScrollTop (element: ElementOrWindow): number {
  return 'scrollTop' in element ? element.scrollTop : element.pageYOffset
}

// 设置滚动高度
export function setScrollTop (element: ElementOrWindow, value: number) {
  'scrollTop' in element ? (element.scrollTop = value) : element.scrollTo(element.scrollX, value)
}

// 获取元素距离顶部高度
export function getElementTop (element: ElementOrWindow) {
  return (
    (element === window ? 0 : (element as HTMLElement).getBoundingClientRect().top) + getScrollTop(window)
  )
}

export function getVisibleHeight (element: ElementOrWindow) {
  return element === window
    ? element.innerHeight
    : (element as HTMLElement).getBoundingClientRect().height
}
