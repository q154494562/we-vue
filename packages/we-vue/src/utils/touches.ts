function getTouch (event: TouchEvent): Touch {
  return event.changedTouches[0] || event.touches[0]
}

export { getTouch }
