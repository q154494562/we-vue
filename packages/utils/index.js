import Vue from 'vue'
import { getTouch } from './touches'

const isServer = Vue.prototype.$isServer

function isDef (value) {
  return value !== undefined && value !== null
}

function isObj (x) {
  const type = typeof x
  return x !== null && (type === 'object' || type === 'function')
}

export {
  isServer,
  getTouch,
  isDef,
  isObj
}
