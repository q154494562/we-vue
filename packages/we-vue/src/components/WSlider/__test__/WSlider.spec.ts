import { mount, shallowMount } from '@vue/test-utils'
import Slider from '../'
import { horizontalDrag } from '@/test/unit/utils'
import faker from 'faker'

describe('slider', () => {
  test('create', () => {
    const wrapper = shallowMount(Slider, {
      propsData: {},
    })

    expect(wrapper.name()).toBe('w-slider')
    expect(wrapper.classes()).toContain('weui-slider-box')

    expect(wrapper.html()).toMatchSnapshot()

    // vue 默认会将异常转为 console.eeror 输出，为此通过 mock console.error 和 console.warn，避免这一机制，方便测试
    jest.spyOn(console, 'error').mockImplementation(() => {})
    jest.spyOn(console, 'warn').mockImplementation(() => {})

    // set the min be bigger than max
    expect(() => {
      shallowMount(Slider, {
        propsData: {
          min: 10,
          max: 1,
        },
      })
    }).toThrow('property:max must be bigger than property:min')
  })

  test('compute percent', () => {
    const fakeValue = faker.random.number({ min: 0, max: 100 })
    const wrapper = shallowMount(Slider, {
      propsData: {
        value: fakeValue,
      },
    })

    const correctPercent = Math.floor((wrapper.vm.value - wrapper.vm.min) / (wrapper.vm.max - wrapper.vm.min) * 100)

    expect(wrapper.vm.percent).toBe(correctPercent)
  })

  test('show-value', () => {
    const wrapper = shallowMount(Slider, {
      propsData: {
        showValue: true,
      },
    })

    expect(wrapper.contains('.weui-slider-box__value')).toBeTruthy()
    wrapper.setProps({
      showValue: false,
    })

    expect(wrapper.contains('.weui-slider-box__value')).toBeFalsy()
  })

  test('disabled', () => {
    const wrapper = shallowMount(Slider, {
      propsData: {
        value: 0,
        disabled: true,
      },
    })

    // click the inner
    wrapper.find({ ref: 'inner' }).trigger('click')

    // drag handler
    horizontalDrag(wrapper.find({ ref: 'handler' }), 0, 10)

    expect(wrapper.vm.value).toBe(0)
  })

  test('drag the handler', () => {
    const wrapper = shallowMount(Slider, {
      propsData: {
        value: 0,
      },
    })

    const { sliderLeft, min, max, step, stepWidth } = wrapper.vm

    const mockClientX = sliderLeft + 50

    horizontalDrag(wrapper.find({ ref: 'handler' }), 0, mockClientX)

    let expectedValue = min + step * (Math.round((mockClientX - sliderLeft) / stepWidth))

    expectedValue = expectedValue < min ? min : expectedValue > max ? max : expectedValue

    expect(wrapper.emitted().input[0]).toEqual([expectedValue])
    expect(wrapper.emitted().change[0]).toEqual([expectedValue])
  })

  test('click the inner', () => {
    const wrapper = mount(Slider, {
      attachToDocument: true,
      propsData: {
        value: 0,
      },
    })

    const { min, step, sliderLeft, stepWidth } = wrapper.vm

    const mockClientX = sliderLeft + 50

    wrapper.find({ ref: 'inner' }).trigger('click', { clientX: mockClientX })

    const expectedValue = min + step * (Math.round((mockClientX - sliderLeft) / stepWidth))

    expect(wrapper.emitted().input[0]).toEqual([expectedValue])
    expect(wrapper.emitted().change[0]).toEqual([expectedValue])
  })
})
