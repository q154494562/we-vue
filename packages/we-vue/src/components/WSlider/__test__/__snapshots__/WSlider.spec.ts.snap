// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slider create 1`] = `
<div class="weui-slider-box">
  <div class="weui-slider">
    <div class="weui-slider__inner">
      <div class="weui-slider__track" style="width: 0%;"></div>
      <div class="weui-slider__handler" style="left: 0%;"></div>
    </div>
  </div>
  <div class="weui-slider-box__value"></div>
</div>
`;
